yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,412 @@
+cimport numpy as np
+import numpy as np
+cimport cython
+
+# for writing to file
+import struct
+
+# cpp includes
+from libcpp.vector cimport vector
+from libcpp cimport bool
+
+# c includes
+cimport libc.math as math
+
+from yt.geometry.particle_deposit cimport \
+    kernel_func, get_kernel_func
+
+cdef struct Node:
+    double left_edge[3]
+    double right_edge[3]  # may be more efficient to store the width instead
+    int start             # which particles we store
+    int end
+    int parent            # position of parent in Octree.nodes
+    int children          # position of 0th child, children are contiguous
+    bool leaf
+    unsigned int node_id  # these probably should be longs
+    unsigned int leaf_id
+    unsigned short depth
+
+cdef struct Octree:
+    vector[Node] nodes
+    double left_edge[3]
+    double right_edge[3]
+    unsigned int n_ref
+    Node * root
+    int * idx
+    unsigned short max_depth
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef class PyOctree:
+    cdef Octree c_tree
+    cdef int[:] _idx
+    cdef int _n_ref
+    cdef int _num_octs
+    cdef int _num_particles
+    cdef kernel_func kernel
+
+    def __init__(self, double[:, ::1] &input_pos = None, left_edge = None,
+                 right_edge = None, int n_ref=32):
+        # if this is the case, we are very likely just initialising an instance
+        # and then going to load an existing kdtree from memory
+        if input_pos is None:
+            return
+
+        self.setup_bounds(input_pos, left_edge, right_edge)
+        self.setup_ctree(input_pos, n_ref)
+        self.setup_root(input_pos)
+
+        # now build the tree, we do this with no gil as we will attempt to
+        # parallelize in the future
+        with nogil:
+            process_node(self.c_tree, self.c_tree.root, &(input_pos[0, 0]))
+
+        # setup the final parameters
+        self.n_ref = n_ref
+        self.num_octs = self.c_tree.nodes.size()
+        self.num_particles = input_pos.shape[0]
+
+    cdef setup_ctree(self, double[:, ::1] &input_pos, int n_ref):
+        self.c_tree.n_ref = n_ref
+        self.c_tree.max_depth = 0
+
+        self.idx = np.arange(0, input_pos.shape[0], dtype=np.int32)
+        self.c_tree.idx = &self._idx[0]
+
+        reserve(&self.c_tree.nodes, 10000000)
```

Comment:
This requires about 1gb of ram. i chose this as a conservative-ish value and if a tree exceeds this size then I'm just relying on the std library to grow. I think the std library grows via doubling capacity, when size hits capacity.

I don't have to worry about pointer invalidation, as i explicitly use vector indices for this exact reason - except my convenience pointer to the root node, which I'll remove.

Reply:
I will maybe mention this in particles/development - I think we can now safely remove the reserve and allow STL to do the re-allocation but this will be slower.